Expose geographies as dimensions.

NeilMadden opened the pull request:
### What

Exposes the geographical hierarchies referenced by a dataset as dimensions to conform with the stub API. Extends the model to cope with hierarchical dimensions. Note: this work does not include any of the dedicated geography endpoints (`/geoareas` etc) in the stub as that is covered by a different ticket.

### How to review

`mvn clean package && java -jar target/dd-*.jar` will run the tests and launch the API. Then load a dataset with hierarchical geography data and check the dimensions.

### Who can review

Anyone apart from me.

NeilMadden commented on src/main/java/uk/co/onsdigital/discovery/metadata/api/dao/MetadataDaoImpl.java:
```diff
@@ -61,4 +64,17 @@ public ConceptSystem findConceptSystemByDataSetAndConceptSystemName(String dataS
+        return dataSet.getReferencedGeographies()
+                .filter(geog -> Objects.equals(geog.getGeographicAreaHierarchy(), geographyId))
+                .findAny()
+                .orElseThrow(() -> new GeographicHierarchyNotFoundException(geographyId));
```
As the comment mentions, this takes the naive approach of iterating through all data points in the dataset. At some point the loader needs to switch from loading the actual data points to just the metadata, at which point it should instead populate a summary table with this information directly.

CarlHembrough commented:
A subtle difference noticed when testing. When getting the dataset, the stub includes the dimensions, but not the options. The API includes all dimension options all the time.

CarlHembrough commented on src/main/java/uk/co/onsdigital/discovery/metadata/api/model/DimensionOption.java:
```diff
@@ -22,9 +30,22 @@ public String getName() {
+                .compare(this.id, that.id, String.CASE_INSENSITIVE_ORDER)
+                .compare(this.name, that.name, String.CASE_INSENSITIVE_ORDER)
+                .compare(this.options, that.options, Ordering.natural().lexicographical())
+                .result();
```
Is it not sufficient to just compare the ID?

NeilMadden replied:
It is for the ordering in the API, this is actually just this complicated so that the unit tests are short and readable. I could extract that into a separate comparator just in the tests I suppose.

CarlHembrough commented on src/main/java/uk/co/onsdigital/discovery/metadata/api/service/MetadataServiceImpl.java:
```diff
@@ -59,14 +62,20 @@ public DataSet findDataSetById(String dataSetId) throws DataSetNotFoundException
-        return convertConceptSystemToDimension(conceptSystem, dataSetId, true);
+        try {
+            final ConceptSystem conceptSystem = metadataDao.findConceptSystemByDataSetAndConceptSystemName(dataSetId, dimensionId);
+            return convertConceptSystemToDimension(conceptSystem, dataSetId, true);
```
We seem to use the terms dimension and concept system interchangeably which could be confusing. Do you think its worth standardising which we use throughout?

NeilMadden replied:
It might be good to push down some Dimension class/interface into the data model. There's really a kind of is-a relationship with both ConceptSystem and GeographicAreaHierarchy being kinds of Dimension.

NeilMadden commented:
I'll check with the front-end team whether the stub or the real API is doing the right thing with the options and fix one or the other.

NeilMadden commented:
I've fixed the API to only return the dimension options on the `.../dimensions` pages and not on the dataset page. Introduced a DimensionBuilder class to make the translation code a bit more readable.
